File: circuit/deposit.go
// Package circuit — DepositCircuit proves that a delta ciphertext correctly
// encrypts a publicly known deposit amount under the account's public key.
//
// Without this proof, anyone could pass a malformed delta_c2 and corrupt
// another account's ciphertext. The proof binds delta_c1 and delta_c2 to a
// consistent randomness R and the public amount.
//
// # Private witnesses
//
//	R – randomness used to construct the delta ciphertext (BN254 Fr)
//
// # Public inputs
//
//	Pk       – account public key (G1)
//	DeltaC1  – delta ciphertext C1 (G1)
//	DeltaC2  – delta ciphertext C2 (G1)
//	Amount   – lamports being deposited (u32, native frontend.Variable)
//
// The circuit proves:
//
//  1. DeltaC1 = R * G
//  2. DeltaC2 = R * Pk + Amount * G
//  3. Amount ∈ [0, 2³²)
package circuit

import (
	"github.com/consensys/gnark-crypto/ecc/bn254"
	"github.com/consensys/gnark/frontend"
	"github.com/consensys/gnark/std/algebra"
	"github.com/consensys/gnark/std/algebra/emulated/sw_bn254"
	"github.com/consensys/gnark/std/math/emulated"
)

// DepositCircuit is the Groth16 deposit constraint system.
type DepositCircuit struct {
	// ── private witnesses ──────────────────────────────────────────────────
	R emulated.Element[sw_bn254.ScalarField] `gnark:",secret"`

	// ── public inputs ──────────────────────────────────────────────────────
	Pk      sw_bn254.G1Affine `gnark:",public"` // account public key
	DeltaC1 sw_bn254.G1Affine `gnark:",public"` // delta C1 = R * G
	DeltaC2 sw_bn254.G1Affine `gnark:",public"` // delta C2 = R * Pk + Amount * G
	Amount  frontend.Variable `gnark:",public"` // lamports deposited
}

func (c *DepositCircuit) Define(api frontend.API) error {
	curve, err := algebra.GetCurve[sw_bn254.ScalarField, sw_bn254.G1Affine](api)
	if err != nil {
		return err
	}
	scalarField, err := emulated.NewField[sw_bn254.ScalarField](api)
	if err != nil {
		return err
	}

	// ── 1. Range check: Amount ∈ [0, 2³²) ────────────────────────────────
	api.ToBinary(c.Amount, 32)

	// ── 2. DeltaC1 = R * G ────────────────────────────────────────────────
	computedC1 := curve.ScalarMulBase(&c.R)
	curve.AssertIsEqual(computedC1, &c.DeltaC1)

	// ── 3. DeltaC2 = R * Pk + Amount * G ─────────────────────────────────
	// Compute Amount*G as (Amount+1)*G − G to avoid ScalarMulBase(0) when
	// Amount = 0.  The scalar Amount+1 ∈ [1, 2³²] is never zero.
	_, _, g1gen, _ := bn254.Generators()
	gPoint := sw_bn254.NewG1Affine(g1gen)
	amtP1Bits := api.ToBinary(api.Add(c.Amount, 1), 33)
	amtP1Scalar := scalarField.FromBits(amtP1Bits...)
	amtP1G := curve.ScalarMulBase(amtP1Scalar)
	rPk := curve.ScalarMul(&c.Pk, &c.R)
	computedC2 := curve.Add(curve.Add(rPk, amtP1G), curve.Neg(&gPoint))
	curve.AssertIsEqual(computedC2, &c.DeltaC2)

	return nil
}

File: circuit/src/deposit.rs
//! Deposit circuit.
//!
//! Proves that a delta ciphertext correctly encrypts a publicly known deposit
//! amount under the account's Baby JubJub public key.
//!
//! # Private witnesses
//!
//! - `r` — randomness used to construct the delta ciphertext
//!
//! # Public inputs (6 native Fr elements)
//!
//! - `pk`       — account BJJ public key   (2 Fr)
//! - `delta_c1` — delta ciphertext C1      (2 Fr)
//! - `delta_c2` — delta ciphertext C2      (2 Fr)
//! - `amount`   — lamports deposited       (1 Fr, native u32)
//!
//! Total: 7 Fr public inputs, IC length = 8.
//!
//! # Constraints
//!
//! 1. `delta_c1 = R * G`
//! 2. `delta_c2 = R * Pk + Amount * G`
//! 3. `Amount ∈ [0, 2³²)`

use ark_bn254::Fr;
use ark_ed_on_bn254::{constraints::EdwardsVar, EdwardsAffine, Fr as BJJFr};
use ark_r1cs_std::{fields::fp::FpVar, prelude::*};
use ark_relations::r1cs::{ConstraintSynthesizer, ConstraintSystemRef, SynthesisError};

use crate::gadgets::{alloc_bjj_scalar_bits, alloc_u32_bits, generator_var, scalar_mul};

#[derive(Clone)]
pub struct DepositCircuit {
    // private
    pub r: Option<BJJFr>,

    // public
    pub pk: Option<EdwardsAffine>,
    pub delta_c1: Option<EdwardsAffine>,
    pub delta_c2: Option<EdwardsAffine>,
    pub amount: Option<u32>,
}

impl ConstraintSynthesizer<Fr> for DepositCircuit {
    fn generate_constraints(self, cs: ConstraintSystemRef<Fr>) -> Result<(), SynthesisError> {
        // ── Public inputs ──────────────────────────────────────────────────────
        let pk_var = EdwardsVar::new_input(cs.clone(), || {
            self.pk.ok_or(SynthesisError::AssignmentMissing)
        })?;
        let c1_var = EdwardsVar::new_input(cs.clone(), || {
            self.delta_c1.ok_or(SynthesisError::AssignmentMissing)
        })?;
        let c2_var = EdwardsVar::new_input(cs.clone(), || {
            self.delta_c2.ok_or(SynthesisError::AssignmentMissing)
        })?;
        // Amount as a native field element
        let amount_fp = FpVar::new_input(cs.clone(), || {
            self.amount
                .map(|a| Fr::from(a as u64))
                .ok_or(SynthesisError::AssignmentMissing)
        })?;

        // ── Private witnesses ──────────────────────────────────────────────────
        let r_bits = alloc_bjj_scalar_bits(cs.clone(), self.r)?;

        // Range check amount ∈ [0, 2^32) by allocating as 32 witness bits and
        // verifying they reconstruct the public amount field element.
        let amount_bits = alloc_u32_bits(cs.clone(), self.amount)?;
        let amount_reconstructed = crate::gadgets::bits_to_fpvar(&amount_bits);
        amount_fp.enforce_equal(&amount_reconstructed)?;

        // ── Constraints ────────────────────────────────────────────────────────
        let g = generator_var();

        // 1. delta_c1 = R * G
        let computed_c1 = scalar_mul(&g, &r_bits)?;
        computed_c1.enforce_equal(&c1_var)?;

        // 2. delta_c2 = R * Pk + Amount * G
        let r_pk = scalar_mul(&pk_var, &r_bits)?;
        let amt_g = scalar_mul(&g, &amount_bits)?;
        let computed_c2 = r_pk + amt_g;
        computed_c2.enforce_equal(&c2_var)?;

        Ok(())
    }
}
